opentelemetry/proto/metrics/v1/metrics.proto

```diff
@@ -371,8 +375,12 @@ message DoubleDataPoint {
   // 1970.
   fixed64 time_unix_nano = 3;
 
-  // value itself.
-  double value = 4;
+  // The value itself.  A point is considered invalid when one of the recognized
+  // value fields is not present inside this oneof.
+  oneof value {
+    double as_double = 4;
+    sfixed64 as_int = 6; 
+  }
```

Comment:
@jmacd was it understood at the time of this PR that this change broke backward compatibility? Specifically the issue that if `v0.7.0` or earlier clients serialize a double value of `0.0` it will appear as a value unset to servers with version `v0.8.0` or later. Non-zero values work fine.

The fact that `double_gauge` and `gauge` use the same tags makes me think this was supposed to be backward compatible. I realize that the metrics proto was not marked stable until `v0.9.0` and these changes were marked as "BREAKING", but up until now I thought that was to signal the compile time breaking of renamed bindings, not actually breaking backward compatibility of the wire format.

Reply:
> was it understood at the time of this PR that this change broke backward compatibility?

It would be fair to say this was poorly understood. We did not consider the impact that it would have on users of 0.7 when we released 0.9. I'm not sure we would have sequenced the changes any differently, but we should have communicated much more strongly about this case. As it happened, the implications only became clear with hindsight, and issues like yours have popped up since the collector and vendors were slow to take up 0.9.

Reply:
@jmacd ok that makes sense. I was just trying to piece together what the expectations were with the breakage here. Is there anywhere we should document this incompatibility so it's easier for others to find?

Reply:
I'm taking the approach that it's safe for now to assume that unset values are metrics with double value `0.0`. I feel the likelihood that clients are sending unset values that should be discarded is low compared with the likelihood that they have an older version of proto. Curious to hear what others think of this.

Reply:
I don't know how we would publish such a notice. We broke an API despite maintaining "wire-level" compatibility. Speaking as a vendor, we advised customers not to update to any 0.9 OTLP exporter until we added backend support and we used version-specific paths in order to maintain 0.7 compatibility. From our perspective, this was still a big mess because the OTel-Go SDK (which we use) and many other SDKs upgraded to 0.9 long before the collector or the vendors were ready with support, which left you and others feeling this pain. Sorry. :(

Reply:
> I'm taking the approach that it's safe for now to assume that unset values are metrics with double value 0.0.

Let me try and diagnose this specifically. Which protocol version is the server parsing as, and which protocol version is the client sending? My guess is that an unset double value (parsed by 0.7) is really a 0.9 integer value.

Reply:
I already tested the issue and here’s my best guess: Protobuf probably doesn’t serialize any data for primitive values of zero since that is their default value anyway, it’s a logical optimization. There’s no need to send anything over the wire. Thus with later versions there is no information for value whatsoever and it appears to be an unset oneof.

Reply:
To be clear I tested serializing v0.7.0 from the client and parsing with v0.9.0 from a server. Note this is my own server implementation, I don’t know what the collector does.

Reply:
> We broke an API despite maintaining "wire-level" compatibility.

I don’t agree that it was wire compatible. But I have a feeling that’s from a misunderstanding about the direction of the version mismatch. I’m talking about newer servers and older clients.

Reply:
> Protobuf probably doesn’t serialize any data for primitive values of zero since that is their default value anyway, it’s a logical optimization.

Yes, although [see the documentation on field presence](https://github.com/protocolbuffers/protobuf/blob/master/docs/field_presence.md) because there are alternate behaviors possible.

We tried to take advantage of a wire-compatible rewrite rule, which says that

```
  double field = 10;
```

can be replaced by

```
  oneof {
    double field = 10;
    int64 alt_field = 11;
  }
```

The problem is that this is only true with explicit field presence, which we didn't realize at the time. We had already called a breaking change, and the breakage happens before the first stable version was declared, so I guess the only thing we could do is go back and revise the changelog to clarify just exactly what breaks. Is that what you're after?

Reply:
> Is that what you're after?

Stepping back a bit, it was a surprise to me to discover that metrics with a double value of zero from a client with proto v0.7.0 would appear [invalid](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L359-L360) to servers using newer proto. I was under the impression that these OTLP changes had been wire-compatible. I would think others will be in the same boat unless this is explicitly communicated or documented. Maybe you feel that's already clear enough? For me explaining exactly what "BREAKING" means would be helpful in the changelog.

Reply:
https://github.com/open-telemetry/opentelemetry-proto/pull/333